### PR TITLE
Use Vector for EventHandler

### PR DESCRIPTION
**Pull request by eatkins**
ArrayList::add is not thread safe. I ran into cases where async tests
using utests would fail even when all of the individual tests passed.
This was because multiple threads called back into the handle method of
the handler instance variable, which just delegated to eventList::add.
When this happened, one of the events would get added to the list as a
null reference, which would manifest as an NPE upstream on the master
process. After this change, my tests stopped failing.

(See the guidelines for contributing, linked above)

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]
lgtm! nice fix, @eatkins!

two thoughts:

1. I think we could merge this into 1.1.x too for 1.1.2
2. would it be possible to construct a test for this? (e.g with async tests in utest)

**Comment by retronym** [comment]
Yep, https://github.com/sbt/sbt/pull/3997 addresses a similar problem in non-forked tests.

**Comment by retronym** [comment]
But I also recently discovered #4010 in the context of Scalacheck forked tests, which has the same symptom you're reporting: a single test failure doesn't fail the entire suite. That wasn't a race condition, but a problem of losing data by clobberring a binding in a map of results, when the test framework created many nested `Task`s for fine grained parallel execution (each test methods in a single class got a parallel `Task`, but because these shared the fully-qualified-name of the enclosing class the result aggregation code only ended up using the status of the final test)

@eatkins How sure are you that this change (alone) fixes the problem you saw?

File: testing/agent/src/main/java/sbt/ForkMain.java [hunk 1]
```diff
@@ -294,7 +295,7 @@ private void runTestTasks(final ExecutorService executor, final Task[] tasks, fi
 					Task[] nestedTasks;
 					final TaskDef taskDef = task.taskDef();
 					try {
-						final List<ForkEvent> eventList = new ArrayList<ForkEvent>();
```
Comment (retronym) [inline testing/agent/src/main/java/sbt/ForkMain.java:297]: Alternatively:

```
final Collection<ForkEvent> eventList = new java.util.concurrent.ConcurrentLinkedDeque<ForkEvent>();
```

Comment (retronym) [inline testing/agent/src/main/java/sbt/ForkMain.java:297]: I agree this is racy.

**Comment by eatkins** [comment]
I added a test to the PR that induces the data race that I was seeing.

@retronym I think my issue was slightly different than #4010. In my case, all of my individual tests passed, but the overall suite was failing.

File: testing/agent/src/main/java/sbt/ForkMain.java (hunk 1 above)
Comment (eatkins) [reply testing/agent/src/main/java/sbt/ForkMain.java:297]: I like this if for no other reason than that it makes it very explicit that we expect that the eventList may be touched by multiple threads.

**Review by eed3si9n — CHANGES_REQUESTED** [review]
LGTM
I'd also vote to rebase this to 1.1.x.

**Comment by eed3si9n** [comment]
It seems https://github.com/sbt/sbt/pull/4030 does just that.

**Comment by dwijnand** [comment]
closing in favour of #4030

**Pull request closed without merging**